### PR TITLE
WIP: Dropdowns

### PR DESCRIPTION
**Pull request by jquense**
local Popper component just for convenience we should move it to react-overlays if it makes sense

### CODE REVIEW

**Comment by jquense** [comment]
I think this is ready. @taion take a look at the Popper stuff pleeease

**Comment by jquense** [comment]
ping

File: V4-breaking-changes.md [hunk 1]
```diff
@@ -34,3 +34,10 @@
 ### NavLink
 
 * add
+
+## Dropdown
+
+* remvoed the `disabled` prop on Dropdown, pass it directly to Dropdown.Toggle
```
Comment (astronomersiva) [inline V4-breaking-changes.md]: nitpick - `removed`

Comment (astronomersiva) [reply V4-breaking-changes.md]: I guess `Toggle` has to be a separate item?

File: src/Popper.js [hunk 2]
```diff
@@ -0,0 +1,196 @@
+import PropTypes from 'prop-types';
```
Comment (taion) [inline src/Popper.js]: oh, weird that the TSBS4 dropdown uses popper... the TWBS3 one didn't, right?

File: src/Dropdown.js [hunk 3]
```diff
@@ -87,69 +77,78 @@ const propTypes = {
...
-  }
-
-  componentWillUpdate(nextProps) {
-    if (!nextProps.open && this.props.open) {
-      this._focusInDropdown = contains(
-        ReactDOM.findDOMNode(this.menu),
-        activeElement(document)
-      );
-    }
+    // this.focusNextOnOpen();
```
Comment (taion) [inline src/Dropdown.js]: ?

File: src/Dropdown.js [hunk 4]
```diff
@@ -159,209 +158,181 @@ class Dropdown extends React.Component {
...
-  focus() {
-    const toggle = ReactDOM.findDOMNode(this.toggle);
+  componentWillUnmount() {
+    if (this.popper) this.popper.destroy();
+  }
 
-    if (toggle && toggle.focus) {
-      toggle.focus();
-    }
+  get hasMenuRole() {
```
Comment (taion) [inline src/Dropdown.js]: why a getter?

File: src/Dropdown.js [hunk 5]
```diff
@@ -159,209 +158,181 @@ class Dropdown extends React.Component {
...
+    if (!this.props.show) return;
     this.toggleOpen(event, eventDetails);
-  }
+  };
 
-  handleKeyDown(event) {
-    if (this.props.disabled) {
+  handleKeyDown = event => {
+    const { key, target } = event;
+    const isInput = /input|textarea/i.test(target.tagName);
```
Comment (taion) [inline src/Dropdown.js]: what an odd way for bootstrap to test for this

File: src/DropdownItem.js [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+import elementType from 'prop-types-extra/lib/elementType';
+
+import chain from './utils/createChainedFunction';
+import SafeAnchor from './SafeAnchor';
+import SelectableContext, { makeEventKey } from './SelectableContext';
+import { createBootstrapComponent } from './ThemeProvider';
+import mapContextToProps from './utils/mapContextToProps';
+import NavContext from './NavContext';
+
+class DropdownItem extends React.Component {
```
Comment (taion) [inline src/DropdownItem.js]: so no more `<MenuItem>`?

File: src/DropdownItem.js [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+    /**
+     * Disable the menu item, making it unselectable.
+     */
+    disabled: PropTypes.bool,
+
+    /**
+     * Styles the menu item as a horizontal rule, providing visual separation between
+     * groups of menu items.
+     */
+    divider: all(
```
Comment (taion) [inline src/DropdownItem.js]: we should probably change this API for dividers

File: src/DropdownItem.js [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+      return (
+        <div
+          role="separator"
+          {...props}
+          className={classNames(className, `${bsPrefix}-divider`)}
+        />
+      );
+    }
+
+    if (header) {
```
Comment (taion) [inline src/DropdownItem.js]: and this

File: src/Dropdown.js [hunk 9]
```diff
@@ -87,69 +84,89 @@ const propTypes = {
...
   constructor(props, context) {
     super(props, context);
 
-    this.handleClick = this.handleClick.bind(this);
-    this.handleKeyDown = this.handleKeyDown.bind(this);
-    this.handleClose = this.handleClose.bind(this);
-
     this._focusInDropdown = false;
     this.lastOpenEventType = null;
+    this.popper = createPopper(this.handleUpdate);
```
Comment (taion) [inline src/Dropdown.js]: should this be deferred to cDM? not sure we should do this if no DOM?

File: src/Dropdown.js [hunk 10]
```diff
@@ -159,209 +176,186 @@ class Dropdown extends React.Component {
...
+      as: Component,
+      alignRight: _0,
+      onSelect: _1,
+      id: _2,
+      onToggle: _3,
+      navbar: _4,
       ...props
     } = this.props;
 
     delete props.onToggle;
```
Comment (taion) [inline src/Dropdown.js]: unnecessary since it's already spread out above

File: src/Dropdown.js [hunk 11]
```diff
@@ -159,209 +176,186 @@ class Dropdown extends React.Component {
...
-    if (dropup) {
-      classes[bsClass] = false;
-      classes.dropup = true;
+    if (this.state.lastShow && !this.props.show) {
+      this._focusInDropdown = this.menu.contains(document.activeElement);
     }
-
-    // This intentionally forwards bsSize and bsStyle (if set) to the
-    // underlying component, to allow it to render size and style variants.
+    // TODO Remove this when next release comes out
```
Comment (taion) [inline src/Dropdown.js]: oh react 16 lifecycles

File: src/DropdownItem.js (hunk 7 above)
Comment (jquense) [reply src/DropdownItem.js]: Cuz it's weird?

Comment (taion) [reply src/DropdownItem.js]: yeah, we were pretty confused by this API :p and couldn't use it for our own dropdown menu dividers!

**Pull request merged**